File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.java
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        final int n = arr.length;
        int[] sortedArray = Arrays.copyOf(arr, n);
        Arrays.sort(sortedArray);
        int rank = 1;
        var numToRankMap = new HashMap<Integer, Integer>();

        for (int i = 0; i < n; i++) {
            if (i > 0 && sortedArray[i] > sortedArray[i-1]) {
                rank++;
            }
            numToRankMap.put(sortedArray[i], rank);
        }

        for (int i = 0; i < n; i++) {
            arr[i] = numToRankMap.get(arr[i]);
        }

        return arr; // TC: O(nlogn), SC: O(n + S); S = sorting algorithm time complexity
    }
}

File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        unordered_map<int, int> numToRankMap;
        set<int> s;
        for (int num : arr) {
            s.insert(num);
        }
        int rank = 1;
        for (int num : s) {
            // cout << num << " is assigned rank: " << rank << endl;
            numToRankMap[num] = rank++;
        }
        for (int i = 0, n = arr.size(); i < n; i++) {
            arr[i] = numToRankMap[arr[i]];
        }

        return arr; // TC: O(nlogn), SC: O(n)
    }
};